File: Tutorials/Codes/Modern_Control_Course/Chapter9/Lesson2/Chapter9_Lesson2.cpp
// Chapter9_Lesson2.cpp
// Eigenvalue-based stability criteria using the Eigen C++ library.
//
// Build example:
//   g++ -std=c++17 Chapter9_Lesson2.cpp -I /path/to/eigen -O2 -o Chapter9_Lesson2
//
// Eigen is header-only and is widely used for state-space computations.

#include <Eigen/Dense>
#include <Eigen/Eigenvalues>
#include <complex>
#include <iostream>
#include <string>
#include <vector>
#include <cmath>

using Matrix = Eigen::MatrixXd;
using Complex = std::complex<double>;

std::string classifyContinuous(const Matrix& A, double tol = 1e-10) {
    Eigen::EigenSolver<Matrix> solver(A);
    auto eigenvalues = solver.eigenvalues();

    double maxReal = -1.0e300;
    for (int i = 0; i < eigenvalues.size(); ++i) {
        maxReal = std::max(maxReal, eigenvalues[i].real());
    }

    if (maxReal < -tol) return "asymptotically stable";
    if (maxReal > tol) return "unstable";

    return "borderline: check semisimplicity of imaginary-axis eigenvalues";
}

std::string classifyDiscrete(const Matrix& A, double tol = 1e-10) {
    Eigen::EigenSolver<Matrix> solver(A);
    auto eigenvalues = solver.eigenvalues();

    double spectralRadius = 0.0;
    for (int i = 0; i < eigenvalues.size(); ++i) {
        spectralRadius = std::max(spectralRadius, std::abs(eigenvalues[i]));
    }

    if (spectralRadius < 1.0 - tol) return "asymptotically stable";
    if (spectralRadius > 1.0 + tol) return "unstable";

    return "borderline: check semisimplicity of unit-circle eigenvalues";
}

void printEigenReport(const std::string& name, const Matrix& A, bool continuous) {
    Eigen::EigenSolver<Matrix> solver(A);

    std::cout << "============================================================\n";
    std::cout << name << "\nA =\n" << A << "\n";
    std::cout << "Eigenvalues:\n";
    for (int i = 0; i < solver.eigenvalues().size(); ++i) {
        std::cout << "  " << solver.eigenvalues()[i] << "\n";
    }

    if (continuous) {
        std::cout << "Continuous-time classification: "
                  << classifyContinuous(A) << "\n";
    } else {
        std::cout << "Discrete-time classification: "
                  << classifyDiscrete(A) << "\n";
    }
}

int main() {
    Matrix A1(2, 2);
    A1 << -1.0, 0.0,
           0.0, -2.0;

    Matrix A2(2, 2);
    A2 << 0.0, -1.0,
           1.0,  0.0;

    Matrix A3(2, 2);
    A3 << 0.0, 1.0,
           0.0, 0.0;

    Matrix Ad1(2, 2);
    Ad1 << 0.8, 0.0,
            0.0, 0.5;

    printEigenReport("Continuous asymptotically stable example", A1, true);
    printEigenReport("Continuous oscillatory borderline example", A2, true);
    printEigenReport("Continuous defective zero-eigenvalue example", A3, true);
    printEigenReport("Discrete asymptotically stable example", Ad1, false);

    return 0;
}

File: Tutorials/Codes/Modern_Control_Course/Chapter9/Lesson2/Chapter9_Lesson2.java
// Chapter9_Lesson2.java
// Eigenvalue-based stability criteria using Apache Commons Math.
//
// Compile/run example:
//   javac -cp commons-math3-3.6.1.jar Chapter9_Lesson2.java
//   java  -cp .;commons-math3-3.6.1.jar Chapter9_Lesson2
//
// On Linux/macOS use ":" instead of ";" in the classpath.

import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.EigenDecomposition;
import org.apache.commons.math3.linear.RealMatrix;

public class Chapter9_Lesson2 {
    static final double TOL = 1e-10;

    public static String classifyContinuous(double[][] data) {
        RealMatrix A = new Array2DRowRealMatrix(data);
        EigenDecomposition eig = new EigenDecomposition(A);

        double maxReal = -Double.MAX_VALUE;
        for (int i = 0; i < A.getRowDimension(); i++) {
            maxReal = Math.max(maxReal, eig.getRealEigenvalue(i));
        }

        if (maxReal < -TOL) return "asymptotically stable";
        if (maxReal > TOL) return "unstable";
        return "borderline: check semisimplicity of imaginary-axis eigenvalues";
    }

    public static String classifyDiscrete(double[][] data) {
        RealMatrix A = new Array2DRowRealMatrix(data);
        EigenDecomposition eig = new EigenDecomposition(A);

        double spectralRadius = 0.0;
        for (int i = 0; i < A.getRowDimension(); i++) {
            double real = eig.getRealEigenvalue(i);
            double imag = eig.getImagEigenvalue(i);
            double magnitude = Math.hypot(real, imag);
            spectralRadius = Math.max(spectralRadius, magnitude);
        }

        if (spectralRadius < 1.0 - TOL) return "asymptotically stable";
        if (spectralRadius > 1.0 + TOL) return "unstable";
        return "borderline: check semisimplicity of unit-circle eigenvalues";
    }

    public static void printReport(String name, double[][] data, boolean continuous) {
        RealMatrix A = new Array2DRowRealMatrix(data);
        EigenDecomposition eig = new EigenDecomposition(A);

        System.out.println("====================================================");
        System.out.println(name);
        System.out.println("Eigenvalues:");
        for (int i = 0; i < A.getRowDimension(); i++) {
            System.out.printf("  %.8f %+ .8fi%n",
                    eig.getRealEigenvalue(i), eig.getImagEigenvalue(i));
        }

        if (continuous) {
            System.out.println("Classification: " + classifyContinuous(data));
        } else {
            System.out.println("Classification: " + classifyDiscrete(data));
        }
    }

    public static void main(String[] args) {
        double[][] A1 = {
            {-1.0, 0.0},
            { 0.0,-2.0}
        };

        double[][] A2 = {
            {0.0, -1.0},
            {1.0,  0.0}
        };

        double[][] A3 = {
            {0.0, 1.0},
            {0.0, 0.0}
        };

        double[][] Ad1 = {
            {0.8, 0.0},
            {0.0, 0.5}
        };

        printReport("Continuous asymptotically stable example", A1, true);
        printReport("Continuous oscillatory borderline example", A2, true);
        printReport("Continuous defective zero-eigenvalue example", A3, true);
        printReport("Discrete asymptotically stable example", Ad1, false);
    }
}
